### PR TITLE
fix(api): precompute tasks-core to KV to eliminate Error 1102 on cold colos

### PR DESCRIPTION
**Pull request by DysektAI**
## Problem

Users hitting the API from cold, low-traffic Cloudflare colos (e.g. CPH) get persistent **Error 1102** on `/api/tarkov/tasks-core`. On a per-colo Cache API MISS, one invocation runs the entire heavy pipeline (3 multi-MB upstream fetches, JSONPath translation walk, adapters, overlay deepMerge, 4.2 MB stringify + `cache.put`). That exceeds the per-invocation CPU/memory ceiling, the 1102 aborts before `cache.put` completes, and the colo stays cold — a self-perpetuating failure loop.

## Fix

Move the heavy transform out of the request path entirely:

- **`scripts/precompute` + `.github/workflows/precompute-tarkov-data.yml`**: a scheduled GitHub Actions run (every 12h, matching `CACHE_TTL_DEFAULT`, plus `workflow_dispatch` with optional `lang`/`gameMode` inputs for targeted busts) executes the existing `createTarkovJsonTasksCoreFetcher` + `applyOverlay` pipeline (imported from `app/server/utils` — not duplicated) and writes each `lang × gameMode` payload to the **`TARKOV_DATA`** KV namespace under `tasks-core-json-v1-{lang}-{gameMode}` via the Cloudflare REST API (7-day `expiration_ttl`). Combinations run sequentially; per-combination failures are recorded and retried next run while the previous KV entry keeps serving.
- **Request path (unchanged since the previous revision)**: `edgeCache` gains a `precomputed` option. `tasks-core` reads the KV blob first and serves it with `X-Cache-Status: PRECOMPUTE`; missing binding/entry, malformed envelope, or KV read failure all fall back to the existing edge-cache path. Operator bypass (`cacheBust`) skips KV as before.
- **`wrangler.toml`**: `TARKOV_DATA` binding (production + `[env.preview]`) now points at the provisioned namespace ID.

### Why GitHub Actions instead of the scheduled Worker (changed from the first revision)

The account is on the **Workers Free tier (10ms CPU/invocation)**. Even a single lang × gameMode combination (JSON.parse of 3 multi-MB payloads, JSONPath walk, adapters, deepMerge, ~4.2MB stringify) far exceeds 10ms CPU, so chunking/self-chaining cannot help; Durable Object alarms and longer CPU limits require Workers Paid, and Queue consumers share the same Worker CPU limit. Supabase Edge Functions were considered but add a second platform + secret surface with their own CPU/wall-clock limits. GitHub Actions runners have no comparable CPU ceiling, the repo already has CI infrastructure, and the write side needs nothing from Cloudflare except the KV REST endpoint. The `workers/tarkov-precompute` Worker from the first revision is **removed** (a stale, undeployable path is just confusion; git history preserves it if the account ever upgrades).

## Human steps to activate (deliberately not done here — read-only CF access)

1. Create a Cloudflare API token scoped to **Account → Workers KV Storage: Edit** (this account only) and set it as the `CLOUDFLARE_API_TOKEN` **repository secret**.
2. Set **repository variables**: `CLOUDFLARE_ACCOUNT_ID` (the account hosting the `tarkovtracker` Pages project) and `TARKOV_DATA_KV_NAMESPACE_ID` = `6034d8d7b7534946bf04110c33ac3b88` (namespace already exists — verified read-only).
3. After merge, trigger the first fill: Actions → "Precompute Tarkov Data" → Run workflow (leave inputs empty for all 32 combinations; full run takes a few minutes).

Until then, behavior is unchanged (KV entry absent → existing edge-cache path).

## Validation

- `npm run typecheck` ✅ (now also typechecks `scripts/precompute` via `tsc -p`)
- `npm run lint` ✅ (zero warnings)
- Targeted tests: `edgeCache.test.ts` (14), `precomputedTarkov.test.ts` (5), `handlers.test.ts` (16), new `scripts/precompute/__tests__/precompute.test.ts` (6) and `kv.test.ts` (4) — 48 passing. New coverage: filter validation, envelope building + 7-day TTL, per-combination failure isolation (pipeline and KV-write failures), REST writer URL/auth/error handling with mocked `fetch`.
- Smoke: `npm run precompute:tarkov` resolves the shared pipeline via tsx tsconfig paths and fails cleanly on missing env/invalid filters.

## Summary by CodeRabbit

* **New Features**
  * Added KV-backed, precomputed “Tarkov” payloads via `edgeCache` (with `PRECOMPUTE` cache status) and automatic fallback to the normal edge cache when precomputed data is missing/invalid.
  * Introduced a scheduled/manual Tarkov precompute pipeline with a new `precompute:tarkov` command, plus supporting CLI utilities and KV REST writer.
* **Bug Fixes**
  * Improved resilience around precomputed reads (missing/malformed/errored entries) and ensured bypass behavior skips the precomputed store.
* **Tests**
  * Expanded coverage for precomputed fast-path, envelope validation, fallback behavior, and bypass handling.
* **Chores / Documentation**
  * Updated project map, Wrangler KV binding (`TARKOV_DATA`), and failure-reporting workflow; added `tsx` tooling.

### CODE REVIEW

**Comment by DysektAI** [comment]
## Review follow-up: findings addressed + Cloudflare readiness verification

### Review findings fixed

- **Bearer scheme case-sensitivity** (codeant-ai, `workers/tarkov-precompute/src/index.ts`): `isAuthorized` now splits the `Authorization` header into scheme and credentials, compares the scheme case-insensitively per RFC 7235 §2.1, and keeps the token comparison exact. Fixed in 708703f6.
- **Invalid filter silently no-ops** (codeant-ai + coderabbitai, `workers/tarkov-precompute/src/precompute.ts`): added `validatePrecomputeFilter`, which checks `lang` against `API_SUPPORTED_LANGUAGES` and `gameMode` against `VALID_GAME_MODES`. `POST /run` now returns **400** with an explicit message listing supported values instead of returning 200 with an empty result set. The cron path is unchanged (no filter). Fixed in 708703f6.
- **Hand-rolled Cloudflare types** (coderabbitai nitpick): `@cloudflare/workers-types` is already a root devDependency, so the ad-hoc `ScheduledController`/`ExecutionContext` aliases were removed and the worker `tsconfig.json` now references the official types (`"types": ["@cloudflare/workers-types"]`). Worker `tsc --noEmit` passes. Fixed in 202c4e4d.

### Additional hardening

- `isPrecomputedEnvelope` now rejects envelopes whose `payload` is `null`/`undefined` (previously `'payload' in candidate` accepted them, which could have served `null` to clients from a corrupt KV write). Falsy-but-valid payloads (`[]`, `0`) still pass. Added edge-case tests: null/undefined payload, `NaN`/`Infinity`/non-numeric `storedAt`, version mismatch. Commit a2443c6e.
- Verified KV key consistency: the handler (`tasks-core.get.ts`) and the Worker both import `buildTasksCorePrecomputedKey` from `app/server/utils/precomputedTarkov.ts`, so the key format cannot drift.
- Worker logging: `console.log`/`console.error` with `[observability] enabled = true` is the correct approach for a standalone Worker (the Nuxt logger is aliased in via wrangler `[alias]` only for the shared `app/server/utils` modules; the shim's empty runtime config is safe since those modules treat it as "use defaults").
- `/run` `summarize` response reviewed: it exposes only duration, counts, KV keys, and upstream error messages behind the bearer-token guard — no secrets.

### Cloudflare verification (read-only, via MCP)

- **Account plan**: the account (DysektAI) is on the Workers **Standard** pricing model but has **no Workers Paid subscription — Workers are on the Free tier**. ⚠️ The Free tier's 10 ms CPU per invocation will not survive the full 34-combination cron run (each combination materializes multi-MB JSON transforms). Before provisioning, either upgrade to Workers Paid (~$5/mo, 30 s CPU default and up to 15 min for cron) or split the schedule into per-lang/per-gameMode chunks via `POST /run` filters.
- **KV namespaces**: none exist on the account — no collision with `TARKOV_DATA`; safe to provision.
- **Pages project**: `tarkovtracker` exists; production branch is `main` as expected.

### CI status

The **Cloudflare Pages** check failure is expected and not a code issue: the `TARKOV_DATA` bindings in `wrangler.toml` use placeholder IDs (`00000000000000000000000000000000`). It will resolve once the namespace is created and the real ID is substituted in both the root `wrangler.toml` (production + preview) and `workers/tarkov-precompute/wrangler.toml`. All other checks pass.

### Validation run locally

- `npm run typecheck` ✅ · `npm run lint` (zero warnings) ✅ · worker `tsc --noEmit` ✅
- `edgeCache.test.ts`, `precomputedTarkov.test.ts`, `handlers.test.ts`: **38/38 passing** ✅

### Suggested follow-up (not in this PR)

- Add unit tests for the precompute Worker itself (`isAuthorized`, `validatePrecomputeFilter`, `runPrecompute` failure isolation) — the Worker currently has no test wiring, and the logic is exercised only indirectly.

**Comment by DysektAI** [comment]
## Follow-up: precompute moved from a scheduled Worker to GitHub Actions

**Why the Worker approach was a dead end**: the account is on the **Workers Free tier — 10ms CPU per invocation**. The tasks-core transform (JSON.parse of 3 multi-MB payloads, JSONPath translation walk, adapters, overlay deepMerge, ~4.2MB stringify) costs well over 10ms CPU for even a *single* lang × gameMode combination, so no amount of chunking (self-chaining, multiple crons, Queues — consumers are Workers with the same limit) fits. Durable Object alarms would allow longer runs but require Workers Paid, which we're avoiding.

**Chosen approach — GitHub Actions** (alternatives considered: Supabase Edge Functions rejected for adding a second platform/secret surface with its own CPU + wall-clock limits; Queue-chunked Workers rejected per above; external cron services rejected as new infrastructure for something CI already does):
- `scripts/precompute/` runs the **same** `createTarkovJsonTasksCoreFetcher` + `applyOverlay` pipeline from `app/server/utils` (zero duplicated transform logic), executed with `tsx` using tsconfig paths + the proven `#imports` shim.
- `.github/workflows/precompute-tarkov-data.yml` runs it every 12h (matching `CACHE_TTL_DEFAULT`) and via `workflow_dispatch` with optional `lang`/`gameMode` inputs (passed as env vars, not shell interpolation — no injection surface). This replaces the Worker's token-guarded `POST /run`; `PRECOMPUTE_TRIGGER_TOKEN` is gone with no remaining references.
- KV writes go through the Cloudflare REST API (`PUT .../values/:key?expiration_ttl=604800`), one request per key — the bulk endpoint can't hold 32 × ~4.2MB in one call, and per-key writes preserve per-combination failure isolation. The 7-day TTL outage buffer is unchanged.

**Changed from the previous PR state**:
- ❌ Removed `workers/tarkov-precompute/` entirely (an undeployable stale path is just confusion; git history preserves it if the account ever upgrades to Paid).
- ➕ Added `scripts/precompute/{run,precompute,kv,nuxt-imports}.ts` + tests (10 new tests, mocked KV/fetch), the Actions workflow, `npm run precompute:tarkov`, and `tsx` as a devDependency (needed to execute the TS script with `@/`/`#imports` path mapping outside Nuxt).
- 🔧 Root `wrangler.toml` placeholders replaced with the real `TARKOV_DATA` namespace ID `6034d8d7b7534946bf04110c33ac3b88` — verified **read-only** that the namespace exists on the same account that hosts the `tarkovtracker` Pages project. `npm run typecheck` now also covers `scripts/precompute`.

**To provision and activate** (nothing was created/deployed from here):
1. Repository **secret** `CLOUDFLARE_API_TOKEN`: Cloudflare API token scoped to *Account → Workers KV Storage: Edit* on this account only.
2. Repository **variables**: `CLOUDFLARE_ACCOUNT_ID`, and `TARKOV_DATA_KV_NAMESPACE_ID` = `6034d8d7b7534946bf04110c33ac3b88`.
3. After merge, run Actions → *Precompute Tarkov Data* once manually for the first fill (~few minutes for all 32 combinations).

**Read path is untouched**: `edgeCache`'s `precomputed` option, the envelope/key format (`tasks-core-json-v1-{lang}-{gameMode}`), the KV binding, and the fallback chain (missing binding/entry/malformed envelope → per-colo Cache API path) are all exactly as reviewed. Until the secret/variables are set, behavior is unchanged.

Validation: `npm run typecheck` ✅ · `npm run lint` ✅ (zero warnings) · 48 targeted tests passing (incl. 10 new for filter validation, envelope + TTL, failure isolation, REST writer errors).

**Comment by DysektAI** [comment]
Addressed the latest CodeRabbit review (2 nitpicks):

- **persist-credentials**: set `persist-credentials: false` on the checkout step in `.github/workflows/precompute-tarkov-data.yml` (2c87ca1d).
- **KV write timeout**: added `AbortSignal.timeout(30_000)` to the KV write `fetch` in `scripts/precompute/kv.ts` so a hung request fails only that key (2c87ca1d).

Also downgraded `tsx` to `^4.22.4` (948f05e4) to comply with the 7-day dependency age policy in AGENTS.md; typecheck, lint, and the precompute tests all pass with the older version.

File: workers/tarkov-precompute/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    failures: result.failures,
+    succeeded: result.successes.length,
+  });
+}
+
+function isAuthorized(request: Request, env: Env): boolean {
+  const token = env.PRECOMPUTE_TRIGGER_TOKEN;
+  if (typeof token !== 'string' || token.length === 0) return false;
+  const header = request.headers.get('authorization') ?? '';
+  return header === `Bearer ${token}`;
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/tarkov-precompute/src/index.ts]: Moot: this finding was fixed in 708703f6 (scheme parsed case-insensitively per RFC 7235 §2.1), and the entire `workers/tarkov-precompute` Worker — including this file and its HTTP `/run` endpoint — was subsequently removed in 1eec77d9 when the precompute moved to a scheduled GitHub Actions workflow (`.github/workflows/precompute-tarkov-data.yml`). There is no bearer-authenticated endpoint left in this PR. Resolving.

File: scripts/precompute/precompute.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+export async function runPrecompute(
+  kv: KvWriter,
+  filter: PrecomputeFilter = {}
+): Promise<PrecomputeResult> {
+  const startedAt = Date.now();
+  const langs = API_SUPPORTED_LANGUAGES.filter((lang) => !filter.lang || lang === filter.lang);
+  const gameModes = VALID_GAME_MODES.filter(
+    (gameMode) => !filter.gameMode || gameMode === filter.gameMode
+  );
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply scripts/precompute/precompute.ts:69]: Fixed: `validatePrecomputeFilter` (added in 708703f6, carried into `scripts/precompute/precompute.ts` in 1eec77d9) rejects any `lang`/`gameMode` value not in `API_SUPPORTED_LANGUAGES`/`VALID_GAME_MODES`, and `scripts/precompute/run.ts` calls it before running and exits non-zero with the error message, so an invalid filter can no longer complete as a silent no-op. Covered by the `validatePrecomputeFilter` tests in `scripts/precompute/__tests__/precompute.test.ts`. Resolving.

[the remark this answers is not in the document]
Comment (DysektAI) [reply scripts/precompute/precompute.ts:69]: Fixed: `validatePrecomputeFilter` (added in 708703f6, carried into `scripts/precompute/precompute.ts` in 1eec77d9) rejects unsupported `lang`/`gameMode` values, and `scripts/precompute/run.ts` validates the filter before running and exits non-zero (`fail(filterError)`), so a typo'd workflow_dispatch input fails the Actions run instead of silently no-opping. Covered by tests in `scripts/precompute/__tests__/precompute.test.ts`. Resolving.

**Comment by DysektAI** [comment]
## Second-opinion production readiness review

**Verdict: READY TO MERGE** ✅ (after CI passes on 2b0a37c0)

### Issues found and fixed
- **Precompute could write a structurally empty payload to KV** (2b0a37c0). The upstream fetcher validates that `data` exists, but a parsing-but-empty upstream response (`tasks: []`) would have been written to KV and served **globally by every colo** until the next successful run — a much larger blast radius than the old per-colo cache. `precomputeTasksCore` now rejects payloads with no tasks; the combination is recorded as a failure and the previous KV entry (7-day TTL) keeps serving. Regression test added.

### Issues found but deferred to follow-ups (non-blocking)
- **PRECOMPUTE/HIT paths re-parse + re-stringify the multi-MB payload** per request (`kv.get(key, 'json')` → Nitro re-serializes). Same cost as the pre-existing HIT path, so no regression; a raw text/stream passthrough would cut CPU further but requires reworking edgeCache's return contract.
- **No retry on KV REST writes** (e.g. transient 429/5xx). Acceptable: the run exits non-zero, the failure lists the exact key, the 12h schedule retries, and the 7-day TTL bridges several failed runs.
- **No KV read timeout** in the request path. Workers KV `get` is runtime-managed and does not hang in practice; errors already fall through to the Cache API path.
- **Silent scheduled-workflow failure for 7+ days** would let KV entries expire and reinstate the cold-miss risk. Consider a failure notification (or a workflow that files an issue) as a follow-up.

### Review threads resolved (3)
- codeant-ai / Bearer case sensitivity (`workers/tarkov-precompute/src/index.ts`) — fixed in 708703f6, then made moot when the Worker was deleted in 1eec77d9.
- codeant-ai / invalid filter silently no-ops (`scripts/precompute/precompute.ts`) — fixed by `validatePrecomputeFilter` + non-zero exit in `run.ts`; verified present in current code with tests.
- coderabbitai / same finding — same fix, resolved.

### Review threads still open
None.

### Whole-project impact assessment
- **Other tarkov routes** (bootstrap, cache-meta, hideout, items, items-lite, map-spawns, prestige, tasks-objectives, tasks-rewards): unaffected. The KV read is gated on `options.precomputed`, which only `tasks-core.get.ts` sets. The `PRECOMPUTE` addition to the header-status union is additive.
- **Cache key**: `buildTasksCorePrecomputedKey('en','regular')` → `tasks-core-json-v1-en-regular`, byte-identical to the old inline template (asserted in `precomputedTarkov.test.ts` and the unchanged expectation in `handlers.test.ts`), so existing edge-cache entries stay valid. Writer and reader both import the same builder — no drift possible.
- **Key coverage**: `getValidatedLanguage` clamps to `API_SUPPORTED_LANGUAGES` (16) and `validateGameMode` to 2 modes — every reachable request key is one of the 32 precomputed combinations.
- **api-gateway Worker**: its routes (`/api/v2/*`, `/api/token*`, `/api/progress*`, `/api/team/progress*`, `api.tarkovtracker.org/*`) never cover `/api/tarkov/*`; no interaction with the new header or KV-backed responses.
- **Frontend**: `useMetadata`/`tarkovCache` consume the JSON body only; nothing parses `X-Cache-Status`. The KV envelope's `payload` is the exact object the fetcher returns, so the response shape is unchanged.
- **Tests/AGENTS.md**: no test asserts a status value that changed; AGENTS.md project map + commands updated per the maintenance contract and accurately describe the mechanism.

### Failure mode analysis (all degrade gracefully)
| Scenario | Behavior |
|---|---|
| KV empty (pre-first-run) | `get` returns `null` → fails `isPrecomputedEnvelope` → falls through to Cache API path (status quo) |
| Binding missing/misconfigured | `getPrecomputedStore` returns `null` → Cache API path |
| Corrupt/legacy envelope (wrong version, null payload, bad `storedAt`) | validator rejects → Cache API path |
| KV read throws | caught, logged as warn → Cache API path |
| Precompute fails 7+ days | entries expire → Cache API path (the original 1102 risk returns; see follow-up on alerting) |
| One bad lang/gameMode combo | recorded as failure, remaining combos still written, run exits non-zero for visibility |
| `cacheBust`/`nocache`/bypass header | KV skipped entirely (`BYPASS`), verified by test |

Headers verified per status: `PRECOMPUTE`/`HIT`/`MISS` get `public, max-age/s-maxage`; `BYPASS`/`DEV`/`STALE` get `no-cache`; overlay meta headers propagate from the precomputed payload's `dataOverlay`.

### Cloudflare verification (read-only)
- KV namespace `6034d8d7b7534946bf04110c33ac3b88` exists, titled **TARKOV_DATA**, currently **0 keys** — expected before merge + first workflow run.
- Pages project **tarkovtracker** exists on the same account; the `wrangler.toml` binding (declared for both production and `[env.preview]`) will attach on the next deploy, same mechanism as the existing `API_GATEWAY_LIMITER` DO binding.
- Only one Worker exists on the account (**api-gateway**); the abandoned `tarkov-precompute` Worker was never left deployed. No other Workers or Pages projects are affected.
- Workflow action pins (`actions/checkout@9c091bb…`, `actions/setup-node@48b55a0…`) verified as real commits in the upstream repos (this workflow has never run, so a bad pin would only have surfaced post-merge).

### Validation run
`npm run typecheck` ✅ (incl. `tsc -p scripts/precompute`) · `npm run lint` ✅ zero warnings · vitest: edgeCache, precomputedTarkov, handlers, scripts/precompute — 49 tests ✅

**Post-merge reminder**: trigger `Precompute Tarkov Data` via workflow_dispatch once after merge so KV is warm before the first 12h cron tick.
